src/node.rs

```diff
@@ -4232,4 +4349,319 @@ mod tests {
+            ShowVMDetails::None,
+            ShowGasDetails::None,
+            false,
+            &system_contracts::Options::BuiltIn,
+        );
+
+        // store the block info with just the tx hash invariant
+        {
+            let mut writer = node.inner.write().unwrap();
+            writer
+                .block_hashes
+                .insert(input_block_number.as_u64(), input_block_hash);
+            writer.blocks.insert(
+                input_block_hash,
+                Block {
+                    transactions: vec![TransactionVariant::Hash(input_tx_hash)],
+                    ..Default::default()
+                },
+            );
+        }
+
+        let actual_tx = node
+            .get_transaction_by_block_number_and_index(
+                BlockNumber::Number(input_block_number),
+                U64::from(0),
+            )
+            .await
+            .expect("failed fetching transaction")
+            .expect("no transaction");
+
+        assert_eq!(input_tx_hash, actual_tx.hash);
+        assert_eq!(Some(U64::from(input_block_number)), actual_tx.block_number);
+    }
+
+    #[tokio::test]
+    async fn test_get_transaction_by_block_number_and_index_fetches_from_fork_if_block_missing() {
+        let mock_server = MockServer::run_with_config(ForkBlockConfig {
+            number: 10,
+            transaction_count: 0,
+            hash: H256::repeat_byte(0xab),
+        });
+        let input_block_hash = H256::repeat_byte(0x01);
+        let input_block_number = U64::from(100);
+        let input_tx_hash = H256::repeat_byte(0x02);
+        mock_server.expect(
+            serde_json::json!({
+                "jsonrpc": "2.0",
+                "id": 0,
+                "method": "eth_getTransactionByBlockNumberAndIndex",
+                "params": [
+                    format!("{:#x}", input_block_number),
+                    "0x1"
+                ],
+            }),
+            TransactionResponseBuilder::new()
+                .set_hash(input_tx_hash)
+                .set_block_hash(input_block_hash)
+                .set_block_number(U64::from(input_block_number))
+                .build(),
+        );
+
+        let node = InMemoryNode::<HttpForkSource>::new(
+            Some(ForkDetails::from_network(&mock_server.url(), None, CacheConfig::None).await),
+            crate::node::ShowCalls::None,
+            ShowStorageLogs::None,
+            ShowVMDetails::None,
+            ShowGasDetails::None,
+            false,
+            &system_contracts::Options::BuiltIn,
+        );
+
+        let actual_tx = node
+            .get_transaction_by_block_number_and_index(
+                BlockNumber::Number(input_block_number),
+                U64::from(1),
+            )
+            .await
+            .expect("failed fetching transaction")
+            .expect("no transaction");
+
+        assert_eq!(input_tx_hash, actual_tx.hash);
+        assert_eq!(Some(U64::from(100)), actual_tx.block_number);
```

Comment:
nit/suggestion: Could be updated to use the variable from earlier
```suggestion
        assert_eq!(Some(input_block_number), actual_tx.block_number);
```